src/ctapipe/image/modifications.py

```diff
@@ -100,6 +111,195 @@ def _smear_psf_randomly(
+    be at least twice the number of waveforms ("nsb_level") from that file
+    that we want to add up. If the NSB file is produced with a level of 25%
+    of dark NSB, and we want to simulate 10x dark NSB, then nsb_level=40 (
+    =10/0.25) and the file must contain at least 80 events. This is to
+    guarantee that the different noise waveforms are not too correlated
+
+    This class can eventually use as NSB file a real data DL0 file which
+    contains pedestal events.
+
+    """
+
+    nsb_file = Path(
+        default_value=None,
+        help="Path to a dedicated NSB-only file (e.g. from sim_telarray)",
+    ).tag(config=True)
+
+    nsb_level = Int(
+        default_value=1,
+        help=(
+            "Number of random instances of the NSB waveforms from the "
+            "NSB file to be added up to the waveform"
+        ),
+    ).tag(config=True)
+
+    n_noise_realizations = Int(
+        default_value=100,
+        help=(
+            "Number of different realizations of the total NSB waveform to "
+            "be created per pixel"
+        ),
+    ).tag(config=True)
+
+    rng_seed = Int(default_value=1, help="Seed for the random number generator").tag(
+        config=True
+    )
+
+    nsb_event_types = List(
+        default_value=[EventType.SKY_PEDESTAL],
+        help="List of event types from which to get the noise waveforms",
+    ).tag(config=True)
+
+    total_noise = dict()
+    # One key per tel_id, each of them is an array of shape
+    # [n_noise_realizations, ngains, npixels, nsamples]
+
+    def __init__(
+        self,
+        subarray: SubarrayDescription,
+        config=None,
+        parent=None,
+        **kwargs,
+    ):
+        """
+
+        Parameters
+        ----------
+        subarray: SubarrayDescription
+            Description of the subarray. Provides information about the
+            camera which are useful in calibration. Also required for
+            configuring the TelescopeParameter traitlets.
+        config: traitlets.loader.Config
+            Configuration specified by config file or cmdline arguments.
+            Used to set traitlet values.
+            This is mutually exclusive with passing a ``parent``.
+        parent: ctapipe.core.Component or ctapipe.core.Tool
+            Parent of this component in the configuration hierarchy,
+            this is mutually exclusive with passing ``config``
+
+        """
+
+        super().__init__(subarray=subarray, config=config, parent=parent, **kwargs)
+        self.rng = np.random.default_rng(self.rng_seed)
+
+        event_type_filter = EventTypeFilter(
+            allowed_types=self.nsb_event_types, parent=self
+        )
+
+        # Read in the waveforms in the NSB-only file. Store in a dictionary
+        # with one key per telescope, containing an array [n_events, n_gains,
+        # n_pixels, n_samples]
+        source = EventSource(input_url=self.nsb_file, skip_calibration_events=False)
```

Comment:
Please use the EventSource as a context manager here, otherwise it remains open indefinitely:
e.g.:
```py3
with EventSource(input_url=self.nsb_file, skip_calibration_events=False) as source:
   ...
```